Implements basic importer for Audacity project files

cofiem opened the pull request:
File extension `.aup`.

Includes deserializing and loading expectations.
Includes basic tests.

Initial step for #1.

cofiem committed:
Implements basic importer for Audacity project files with extension .aup.

Includes deserializing and loading expectations.
Includes basic tests.

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+            IEnumerable<string> resultList = results.ToList();
+            return resultList.Any() && resultList.All(p => Path.GetExtension(p) == ProjectFileExtension)
+                ? Some(resultList)
+                : None;
```
```suggestion
            if (errorList.Any())
            {
                return errorList.ToSeq();
            }
            
            return results.Any() && results.All(p => Path.GetExtension(p) == ProjectFileExtension)
                ? Some(results)
                : None;
```

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+        {
+            if (context.Include.Filter is not null)
+            {
+                throw new Exception("The Audacity importer does not currently support include filters");
```
There's nothing useful to filter on in avianz data files, but you could filter the different audacity tracks by name?

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+                                    Name = name,
+                                };
+                        }))
+                    .ToList() ?? new List<IExpectation> {new NoEvents()};
```
I'd prefer to avoid the allocation of the list here

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+                    logger.LogWarning("No annotations found in {path}, producing a no events expectation", path);
+                }
+
+                if (context.Include.Exhaustive.HasValue && context.Include.Exhaustive.Value)
```
```suggestion
                if (context.Include.Exhaustive is bool exhaustive))
```

Without this change the behaviour is not equivalent to the avianz importer. Why?

Because the config normalization does this later?

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+
+                logger.LogTrace("Data file converted to expectations: {@expectations}", expectations);
+
+                yield return new TestCase {SourceInfo = dataFile?.SourceInfo, Expect = expectations.ToArray(),};
```
dataFile should never be null right? If it is, then we should fall back to providing some source info (like the `path`) so in error cases at least we can print a message that point to the right source  file.

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacitySerializer.cs:
```diff
@@ -0,0 +1,37 @@
+
+        public Project Deserialize(Stream stream, string path)
+        {
+            var result = serializer.Deserialize<Project>(stream);
```
clarify whether this can be `null`. I don't think it should be (if the file is empty it should be an empty `Project` instance

atruskie commented on src/Egret.Cli/Serialization/Audacity/LabelTrack.cs:
```diff
@@ -0,0 +1,39 @@
+        public List<Label> Labels { get; init; }
+
+        [XmlAttribute(AttributeName = "numlabels")]
+        public int NumLabels { get => Labels.Count; init => _ = value; }
```
I assume discard instead of throw/not have setter to make the serializer happy?

atruskie commented on src/Egret.Cli/Serialization/Audacity/Project.cs:
```diff
@@ -0,0 +1,68 @@
+        public double Rate { get; init; } = 44100.0;
+
+        [XmlAttribute(AttributeName = "version")]
+        public string Version { get; init; } = "1.3.0";
+
+        [XmlAttribute(AttributeName = "audacityversion")]
+        public string AudacityVersion { get; init; } = "2.4.2";
```
I'm a little sceptical that these should be default values.

Can audacity load an .aup file if these values are missing? If yes, case remove defautls.

atruskie commented on src/Egret.Cli/Serialization/Audacity/Tag.cs:
```diff
@@ -0,0 +1,23 @@
+        public string Name { get; init; }
+
+        [XmlAttribute(AttributeName = "value")]
+        public string Value { get; init; }
```
~~What does the `Value` attribute do? Just nice to know~~

It'd be good to leave a note that a `Tag` actually acts like a key value pair

atruskie commented on src/Egret.Cli/Serialization/Xml/XmlSerializer.cs:
```diff
@@ -0,0 +1,85 @@
+    using System.Xml;
+    using System.Xml.Serialization;
+
+    public class DefaultXmlSerializer
+    {
+        public Action<object, XmlAttributeEventArgs> OnUnknownAttribute { get; set; }
+
+        public Action<object, XmlElementEventArgs> OnUnknownElement { get; set; }
+
+        public Action<object, XmlNodeEventArgs> OnUnknownNode { get; set; }
+
+        public Action<object, UnreferencedObjectEventArgs> OnUnreferencedObject { get; set; }
+        
+        public async Task<T> Deserialize<T>(IFileInfo fileInfo)
```
The way these event bindings are done this class can not be a singleton in the DI service

..I just looked, it's not even there?

It should be added as a transient in DI service

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityImporterTests.cs:
```diff
@@ -0,0 +1,193 @@
+
+            result.Tracks.Should().HaveCount(2);
+
+            result.Tracks.First().Should().Match<LabelTrack>(i =>
+                i.Name == "Track 2" &&
+                i.IsSelected == 1 &&
+                i.Height == 206 &&
+                i.Minimized == 0 &&
+                i.Labels.Count == 2
+            );
+            result.Tracks.First().Labels.First().Should().Match<Label>(i =>
+                i.Title == "test 1" &&
+                i.TimeStart.AlmostEqual(4) &&
+                i.TimeEnd.AlmostEqual(7.2446258503) &&
+                i.SelLow.AlmostEqual(1) &&
+                i.SelHigh.AlmostEqual(10)
+            );
+            result.Tracks.First().Labels.Skip(1).First().Should().Match<Label>(i =>
+                i.Title == "test 3" &&
+                i.TimeStart.AlmostEqual(15.9714285714) &&
+                i.TimeEnd.AlmostEqual(24.44) &&
+                i.SelLow.AlmostEqual(10) &&
+                i.SelHigh.AlmostEqual(10000)
+            );
+
+            result.Tracks.Skip(1).First().Should().Match<LabelTrack>(i =>
+                i.Name == "Track 1" &&
+                i.IsSelected == 1 &&
+                i.Height == 90 &&
+                i.Minimized == 0 &&
+                i.Labels.Count == 1
+            );
+            result.Tracks.Skip(1).First().Labels.First().Should().Match<Label>(i =>
+                i.Title == "test 2" &&
+                i.TimeStart.AlmostEqual(8.0228571429) &&
+                i.TimeEnd.AlmostEqual(18.98) &&
+                i.SelLow == 0 &&
+                i.SelHigh == 0
+            );
+        }
```
break this out into a test file for the audacity serializer please

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityImporterTests.cs:
```diff
@@ -0,0 +1,193 @@
+test_suites:
+  host_suite:
+    tests:
+      - file: bird*.wav
```
this line of the config should be redundant

in an abstract sense: the main egret config is importing other tests (e.g. from another egret file, or an aup file acting like an egret file) and those tests should be referencing audio files.

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityImporterTests.cs:
```diff
@@ -0,0 +1,193 @@
+                        break;
+                    case 2:
+                        e.Name.Should().Be($"Audacity Label test 2 in track Track 1 (1:0)");
+                        break;
+                }
+
+                switch (e)
+                {
+                    case BoundedExpectation bounded:
+                        switch (index)
+                        {
+                            case 0:
+                                bounded.Bounds.StartSeconds.Should().Be(4.0.WithTolerance(temporalTolerance));
+                                bounded.Bounds.EndSeconds.Should().Be(7.2446258503.WithTolerance(temporalTolerance));
+                                bounded.Bounds.HighHertz.Should().Be(10.0.WithTolerance(spectralTolerance));
+                                bounded.Bounds.LowHertz.Should().Be(1.0.WithTolerance(spectralTolerance));
+                                break;
+                            case 1:
+                                bounded.Bounds.StartSeconds.Should().Be(15.9714285714.WithTolerance(temporalTolerance));
+                                bounded.Bounds.EndSeconds.Should().Be(24.44.WithTolerance(temporalTolerance));
+                                bounded.Bounds.HighHertz.Should().Be(10000.0.WithTolerance(spectralTolerance));
+                                bounded.Bounds.LowHertz.Should().Be(10.0.WithTolerance(spectralTolerance));
+                                break;
+                            default:
+                                throw new XunitException();
+                        }
+                        break;
+                    case TemporalExpectation temporal:
+                        switch (index)
+                        {
+                            case 2:
+                                temporal.Time.StartSeconds.Should().Be(8.0228571429.WithTolerance(temporalTolerance));
+                                temporal.Time.EndSeconds.Should().Be(18.98.WithTolerance(temporalTolerance));
+                                break;
+                            default:
+                                throw new XunitException();
+                        }
+                        break;
+                }
+            });
```
I would probably turn this into a static structure rather than use multiple switch statements

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityImporterTests.cs:
```diff
@@ -0,0 +1,193 @@
+            var temporalTolerance = includeTests.TemporalTolerance ?? 0.5;
+            var spectralTolerance = includeTests.SpectralTolerance ?? 0.5;
+
+            tests.ForEach((e, index) =>
```
```suggestion
            expectations.ForEach((e, index) =>
```

suite: many files
test: one file + some expectations
expectation: one event that we expect with some assertions
assertion: checking one individual attribute of the event

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityImporterTests.cs:
```diff
@@ -0,0 +1,193 @@
+
+            var tests = testsCases[0].Expect;
+
+            tests[0].Should().BeOfType<BoundedExpectation>();
```
The generated test should have it's `File` property set, test that

atruskie commented on src/Egret.Cli/Serialization/Audacity/AudacityImporter.cs:
```diff
@@ -0,0 +1,118 @@
+
+                logger.LogTrace("Data file converted to expectations: {@expectations}", expectations);
+
+                yield return new TestCase {SourceInfo = dataFile?.SourceInfo, Expect = expectations.ToArray(),};
```
The `File` property is not set - it must be set. The assumption here is that the importer knows which file the metadata is referring too.

Note to self: should probably make TestCase throw if both File and Url are null or empty

atruskie replied:
As in https://github.com/QutEcoacoustics/egret/blob/c01f5111fb05813fbe6413d5a4cd56a06b5c822d/src/Egret.Cli/Serialization/Avianz/AvianzImporter.cs#L85

atruskie reviewed (changes requested):
So confirm for me: should there be one audacity project file per audio file?
And are those project files required to have the same name as the audio file?

If not, then we'll need to rethink the structure of this a bit. If so, see comments

Also I'm going to guess you didn't run the app? Unless I'm mistaken not enough of the new services were registered in the DI system.

Otherwise, excellent 😁

cofiem committed:
Changes in response to feedback, plus bonus Audacity 3 project implementation.

Registered relevant classes as services.
Improved tests.
Made XmlSerializer work as a service.

atruskie commented on docs/audacity.md:
```diff
@@ -0,0 +1,105 @@
+The directory contains the audio imported into Audacity. You may want to keep the directory if you want to change the labels.
+
+Make sure you keep the original audio file you imported into Audacity.
+Egret requires this original audio file to run the analysis tools and get results.
```
Add a note on how we resolve the original file

atruskie commented on docs/audacity.md:
```diff
@@ -0,0 +1,105 @@
+
+## <a id="usageReviewResults"></a> Usage for reviewing results
+
+When running Egret, there is an option to save the results as Audacity Project files using `--audacity`.
```
```suggestion
When running Egret, there is an option to save the results as Audacity Project files using the `--audacity` command line argument.
```

atruskie commented on docs/audacity.md:
```diff
@@ -0,0 +1,105 @@
+
+## QUESTION
+
+Change to using audacity files as expectations, without specifying the file to test?
```
```suggestion
Can I change to using audacity files as expectations, without specifying the file to test?
```

atruskie commented on docs/audacity.md:
```diff
@@ -0,0 +1,105 @@
+When saving, it likely does not make sense to try to build the audacity project audio file structure (need to process and split files),
+as Egret will need to be able to process audio if saving audio in the Audacity format.
+
+Maybe use the project tags to reference one file per .aup output file?
```
I think we can simplify this section a little bit - see overall comments.

atruskie commented on src/Egret.Cli/Egret.Cli.csproj:
```diff
@@ -24,9 +24,13 @@
-    <PackageReference Include="Microsoft.Extensions.Logging" Version="3.1.9" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="5.0.4" />
+
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite.Design" Version="1.1.6" />
```
I'm probably going to veto EntityFramework - that's a very large API surface for what amounts to a few sqlite3 queries

atruskie commented on src/Egret.Cli/Formatters/AudacityResultFormatter.cs:
```diff
@@ -0,0 +1,110 @@
+            var outputName = outputBaseFile.Filestem();
+
+            var outputFullName = outputName + "_" + sourceName;
+            var outputFullFile = new FileInfo(Path.Combine(outputBaseFile.DirectoryName, outputFullName + ".aup"));
```
Add an overload to `GetOutputFile` that can accept tags to add to the filename

```suggestion
            var sourceName = Path.GetFileNameWithoutExtension(result.Context.SourceName);

            var outputFile = outputFile.GetOutputFile(".aup", sourceName);
```

atruskie commented on src/Egret.Cli/Formatters/AudacityResultFormatter.cs:
```diff
@@ -0,0 +1,110 @@
+        private (FileInfo, Project) BuildProject(int index, TestCaseResult result)
+        {
+            var sourceFile = new FileInfo(result.Context.SourceName);
+            var sourceName = sourceFile.Filestem();
```
Would recommend using `Path.GetFileNameWithoutExtension` here and not converting to FileInfo. Currently SourceName only points to local files but a future feature of egret will be to support remote file - so a URI might be in this spot.

atruskie commented on src/Egret.Cli/Formatters/AudacityResultFormatter.cs:
```diff
@@ -0,0 +1,110 @@
+        private readonly AudacitySerializer serializer;
+        private readonly List<FileInfo> outputFiles;
+
+        public IReadOnlyCollection<FileInfo> OutputFiles => new ReadOnlyCollection<FileInfo>(outputFiles);
```
This list is only written to in this class. What is it for?

atruskie commented on src/Egret.Cli/Formatters/AudacityResultFormatter.cs:
```diff
@@ -0,0 +1,110 @@
+                new Tag(nameof(result.Context.CaseTracker), result.Context.CaseTracker.ToString()),
+                new Tag(nameof(result.Context.ExecutionIndex), result.Context.ExecutionIndex.ToString()),
+                new Tag(nameof(result.Context.ExecutionTime), result.Context.ExecutionTime.ToString()),
+                new Tag(nameof(result.Context.ExecutionTime), result.Context.ExecutionTime.ToString()),
```
Duplicate tag

atruskie commented on src/Egret.Cli/Models/Audacity/Project.cs:
```diff
@@ -1,27 +1,24 @@
 {
     using Models;
-    using System.Collections.Generic;
+    using Serialization.Yaml;
```
Extra namespace?

atruskie commented on src/Egret.Cli/Models/Audacity/Tag.cs:
```diff
@@ -0,0 +1,38 @@
+            Value = value;
+        }
+
+        public static implicit operator KeyValuePair<string, string>(Tag tag)
+        {
+            return new(tag.Name, tag.Value);
+        }
```
If you want to get super fancy you can define an implicit cast from a tuple. It can make for some pretty slick syntax.

```
tags.Add(("key", "value"));
```

atruskie commented on src/Egret.Cli/Program.cs:
```diff
@@ -51,6 +51,7 @@ private static CommandLineBuilder BuildCommandLine()
                     new Option<bool>("--csv", description: "Output results to a CSV file").WithAlias("-c"),
                     new Option<bool>("--no-console", description: "Do not output results in the console").WithAlias("-q"),
                     new Option<bool>("--html", description: "Output results to a HTML file").WithAlias("-h"),
+                    new Option<bool>("--audacity", description: "Output results to Audacity project files").WithAlias("-aud"),
```
I believe System.CommandLine supports short flags option combining. In that case I think multi letter short form is probably confusing. 

e.g. `-chaud` is valid?

Just use `-a` - I assume you didn't want to use up a flag?

atruskie commented on src/Egret.Cli/Serialization/Audacity/Audacity3Importer.cs:
```diff
@@ -0,0 +1,142 @@
+                    
+                    // TODO: audacity projects can store audio files with the project file,
+                    //       but the audio files are split into multiple smaller files.
+                    // File = Path.GetRelativePath(Path.GetDirectoryName(path), testFile)
```
See summary comments - we should find a source file with a similar name next to the project file, or fail

atruskie commented on src/Egret.Cli/Serialization/Audacity/Audacity3Importer.cs:
```diff
@@ -0,0 +1,142 @@
+        private IExpectation MakeNoEventsExpectation(string path)
+        {
+            logger.LogWarning("No annotations found in {path}, producing a no_events expectation", path);
+            return new NoEvents();
```
Unused?

atruskie commented on tests/Egret.Tests/Egret.Tests.csproj:
```diff
@@ -24,7 +24,13 @@
+      <None Update="Serialization\Audacity\Examples\audacity-example2.aup">
+        <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+      </None>
+      <None Update="Serialization\Audacity\Examples\audacity-example2.aup3">
```
```suggestion
      <None Update="..\Fixtures\Audacity\*">
```

atruskie commented on tests/Egret.Tests/Serialization/Audacity/AudacityTests.cs:
```diff
@@ -0,0 +1,131 @@
+            var includeTests = testSuit.IncludeTests[0];
+            includeTests.From.Should().Be(@"/abc/example1.aup");
+
+            var testsCases = includeTests.Tests;
+            testsCases.ToList().Should().HaveCount(1);
+
+            var expectations = testsCases[0].Expect;
```
should assert test's source is set correctly

atruskie reviewed (changes requested):
Since there seems to be no strong link between audacity projects and audio files we're going to enforce a convention.

The file `/123/abc.wav` is related to the file `/123/abc.aup` or `/123/abc.aup3` IFF the directory and basename are equivalent (i.e. they are siblings). When outputting results a source of `/123/abc.wav` will produce `/output/abc.aup`.

It appears the formatter already does this.

I think you should remove the entity framework package. The audacity3 serializer even only uses `Microsoft.Data.Sqlite` which AFAIK is its own package.

Probably only need one audacity importer (it can switch which serializer it uses based on file extension). As far as I can tell, the two importers do the same thing?

cofiem committed:
Functional implementation of Audacity reading and writing.

Addressed feedback - audio file convention, single importer, moved fixture data.
Partial implementation of audacity result formatter.

cofiem commented:
I'm now able to run egret with audacity project files as input, and get audacity files as output!

I'm not sure about how to write the results to the Audacity Project Tracks and labels - there's a few TODOs in there that will need your help @atruskie.

I've made the second round of changes to address your feedback. Ready for another review.
